library/proc_macro/src/bridge/server.rs

```diff
@@ -143,6 +144,33 @@ pub trait ExecutionStrategy {
     ) -> Buffer;
 }
 
+thread_local! {
+    /// While running a proc-macro with the same-thread executor, this flag will
+    /// be set, forcing nested proc-macro invocations (e.g. due to
+    /// `TokenStream::expand_expr`) to be run using a cross-thread executor.
+    ///
+    /// This is required as the thread-local state in the proc_macro client does
+    /// not handle being re-entered, and will invalidate all `Symbol`s when
+    /// entering a nested macro.
+    static RUNNING_SAME_THREAD: Cell<bool> = Cell::new(false);
+}
+
+struct RunningSameThread(());
+
+impl RunningSameThread {
+    fn new() -> Self {
+        let prev = RUNNING_SAME_THREAD.replace(true);
+        assert!(!prev, "cannot run same-thread client recursively");
```

Comment:
```suggestion
        let already_running = ALREADY_RUNNING_SAME_THREAD.replace(true);
        assert!(!already_running, "same-thread nesting ("reentrance") of proc macro executions is not supported");
```
(part of my reasoning is that `assert!($cond` will print `$cond` on failure, even when a message is *also* provided, so a more informative name could make the failure more readable)

Alternatively, if/since you are going to reformat that on multiple lines, you could end up with the `replace` call itself in the `assert!`, if you're okay with that, heh.